### PR TITLE
[core] redesign subscriber api

### PR DESCRIPTION
**Pull request by rex-schilasky**
### Description

**Changes Summary**:
1. **Namespace Updates**:
   - The legacy `eCAL::CSubscriber` API has been moved to the `eCAL::v5` namespace for compatibility with existing systems using eCAL version 5.
   - Introduced a new, modern `eCAL::CSubscriber` API under the `eCAL::v6` namespace (default inline namespace).

2. **API Modernization**:
   - **Initialization**:
     - Removed `Create` and `Destroy` methods in `v6`. Subscribers are now fully initialized via constructors and cleaned up automatically via destructors.
   - **Callback Registration**:
     - Event callbacks (`AddEventCallback`, `RemEventCallback`) are now passed as optional parameters during object construction, simplifying event handling.
     - Receive callbacks renamed `AddReceiveCallback` to `SetReceiveCallback`, `RemReceiveCallback` to `RemoveReceiveCallback`
     - `SetReceiveCallback` and `RemoveReceiveCallback` methods remain part of the API for flexibility in managing message reception callbacks.
   - **Removed Legacy Methods**:
     - Methods such as `SetAttribute`, `ClearAttribute`, and `SetID` were removed as they were experimental or rarely used.
   - **Streamlined Methods**:
     - The `ReceiveBuffer` method in `v5` was replaced with `SetReceiveCallback` to encourage a more event-driven approach to data reception in `v6`.

3. **Code Refactoring**:
   - Internally uses `std::shared_ptr<CSubscriberImpl>` for robust resource management and memory safety.
   - Constructors support optional `SDataTypeInformation` for specifying topic metadata and a `SubEventIDCallbackT` for event-driven designs.

4. **Improved Usability**:
   - Constructors now provide better defaults and more configuration options, making the API easier to use in a wide range of scenarios.
   - Cleaner separation of concerns and removal of outdated features.

**Impact**:
- Maintains backward compatibility with `eCAL::v5::CSubscriber` while offering a modern, streamlined API for new applications.
- Improves code readability, usability, and performance for developers integrating eCAL.

### CODE REVIEW

File: ecal/core/include/ecal/msg/dynamic.h [hunk 1]
```diff
@@ -68,13 +68,13 @@
...
   template <typename T, typename DynamicDeserializer>
-  class CDynamicMessageSubscriber final : public CSubscriber
+  class CDynamicMessageSubscriber final : public v5::CSubscriber
   {
   public:
     /**
      * @brief  Constructor.
     **/
-    CDynamicMessageSubscriber() : CSubscriber()
+    CDynamicMessageSubscriber() : v5::CSubscriber()
```
[the remark this answers is not in the document]
Comment (rex-schilasky) [reply ecal/core/include/ecal/msg/dynamic.h:77]: fixed in next commit

File: ecal/core/include/ecal/msg/dynamic.h [hunk 2]
```diff
@@ -68,13 +68,13 @@ namespace eCAL
...
```
[the remark this answers is not in the document]
Comment (rex-schilasky) [reply ecal/core/include/ecal/msg/dynamic.h:77]: Fixed in next commit

File: ecal/core/include/ecal/msg/subscriber.h [hunk 3]
```diff
@@ -44,13 +44,13 @@
...
   template <typename T>
-  class [[deprecated("Please use CMessageSubscriber instead")]] CMsgSubscriber : public CSubscriber
+  class [[deprecated("Please use CMessageSubscriber instead")]] CMsgSubscriber : public v5::CSubscriber
   {
   public:
     /**
      * @brief  Constructor.
     **/
-    CMsgSubscriber() : CSubscriber()
+    CMsgSubscriber() : v5::CSubscriber()
```
[the remark this answers is not in the document]
Comment (rex-schilasky) [reply ecal/core/include/ecal/msg/subscriber.h:53]: Fixed in next commit

File: ecal/core/include/ecal/msg/subscriber.h [hunk 4]
```diff
@@ -240,13 +240,13 @@
...
   template <typename T, typename Deserializer>
-  class CMessageSubscriber final : public CSubscriber
+  class CMessageSubscriber final : public v5::CSubscriber
   {
   public:
     /**
      * @brief  Constructor.
     **/
-    CMessageSubscriber() : CSubscriber()
+    CMessageSubscriber() : v5::CSubscriber()
```
[the remark this answers is not in the document]
Comment (rex-schilasky) [reply ecal/core/include/ecal/msg/subscriber.h:249]: Fixed with next commit

**Review by KerstinKeller — APPROVED** [review]

File: ecal/core/src/pubsub/ecal_subscriber.cpp [hunk 5]
```diff
@@ -24,199 +24,85 @@
...
 
-    rhs.m_datareader = nullptr;
-
-    return *this;
+    // register subscriber
+    if (g_subgate() != nullptr) g_subgate()->Register(topic_name_, m_subscriber_impl);
   }
 
-  bool CSubscriber::Create(const std::string& topic_name_, const SDataTypeInformation& data_type_info_, const Subscriber::Configuration& config_)
+  CSubscriber::CSubscriber(const std::string& topic_name_, const SDataTypeInformation& data_type_info_, const SubEventIDCallbackT event_callback_, const Subscriber::Configuration& config_) :
```
[the remark this answers is not in the document]
Comment (KerstinKeller) [reply ecal/core/src/pubsub/ecal_subscriber.cpp:43]: Rex: did you want to turn off this warning? doesn't it make sense in this case? Why not a reference?

**Pull request merged**